ios/engine/KMEI/KeymanEngine/Classes/Manager.swift

```diff
@@ -0,0 +1,3018 @@
+        kmLog("Language Directory contents: \(String(describing: contents))", checkDebugPrinting: true)
+      }
+    }
+  }
+
+  public func downloadRequestStarted(_ request: HTTPDownloadRequest) {
+    // If we're downloading a new keyboard.
+    // The extra check is there to filter out other potential request types in the future.
+    if request.tag == 0 && request.typeCode == .downloadFile {
+      NotificationCenter.default.post(name: .keymanKeyboardDownloadStarted, object: self, userInfo: request.userInfo)
+    }
+  }
+
+  public func downloadRequestFinished(_ request: HTTPDownloadRequest) {
+    switch request.typeCode {
+    case .downloadFile:
+      let kbInfo = request.userInfo[Key.keyboardInfo]
+      let keyboard: [String: String]
+      if let kb = kbInfo as? [String: String] {
+        keyboard = kb
+      } else if let keyboards = kbInfo as? [[String: String]], !keyboards.isEmpty {
+        keyboard = keyboards[0]
+      } else {
+        kmLog("downloadRequestFinished(): Unexpected userInfo: \(String(describing: kbInfo))",
+          checkDebugPrinting: false)
+        return
+      }
+      let kbID = keyboard[Key.keyboardId]!
+      let kbVersion = keyboard[Key.keyboardVersion]
+      let isUpdate = (request.userInfo[Key.update] as? Int)! != 0
+
+      if let statusCode = request.responseStatusCode, statusCode == 200 {
+        // The request has succeeded.
+        if downloadQueue!.requestsCount == 0 {
+          // Download queue finished.
+          downloadQueue = nil
+          registerCustomFonts()
+          kmLog("Downloaded keyboard: \(kbID).", checkDebugPrinting: true)
+
+          NotificationCenter.default.post(name: .keymanKeyboardDownloadCompleted, object: self,
+                                          userInfo: request.userInfo)
+          if isUpdate {
+            shouldReloadKeyboard = true
+            reloadKeyboard()
+          }
+          let userData = activeUserDefaults()
+          userData.set([Date()], forKey: Key.synchronizeSWKeyboard)
+          userData.synchronize()
+        }
+      } else { // Possible request error (400 Bad Request, 404 Not Found, etc.)
+        downloadQueue!.cancelAllOperations()
+        downloadQueue = nil
+
+        let errorMessage = "\(request.responseStatusMessage ?? ""): \(request.url)"
+        let error = NSError(domain: "Keyman", code: 0,
+                            userInfo: [NSLocalizedDescriptionKey: errorMessage])
+        kmLog("Keyboard download failed: \(error).", checkDebugPrinting: true)
+
+        if !isUpdate {
+          let fileName = request.url.lastPathComponent
+          if fileName.hasJavaScriptExtension {
+            if let kbPath = keyboardPath(forFilename: fileName, keyboardVersion: kbVersion) {
+              try? FileManager.default.removeItem(at: kbPath)
+            }
+          } else if fileName.hasFontExtension {
+            // TODO: Why do we delete a keyboard with that name?
```

Comment:
In the case of a failed keyboard download, it appears that we pre-emptively started adding it to our managed database but failed to actually retrieve the keyboard.  We should purge all evidence of the download so that we don't later mistakenly believe the download was a success, just to find empty files.  (Unless it can be made clear that we only set said database info upon a successful load, which is clearly not something triggered in the 'success' part of this function.)

I assume we trigger it with the font so that a keyboard is viewed as a single unit with its required fonts; all or nothing.

Note - this section is conditioned upon the `isUpdate` variable - if we were updating, that doesn't render the original keyboard unusable.